File: algorithms/02_c/strings/kmp.c
#include <stdio.h>
#include <string.h>
void compute_lps(const char *p, int m, int *lps){
    int k=0; lps[0]=0;
    for(int i=1;i<m;){
        if(p[i]==p[k]) lps[i++]=++k;
        else if(k) k=lps[k-1];
        else lps[i++]=0;
    }
}
void kmp(const char *text, const char *pat){
    int n=strlen(text), m=strlen(pat), lps[m];
    compute_lps(pat,m,lps);
    int i=0,j=0;
    while(i<n){
        if(text[i]==pat[j]){i++;j++;}
        if(j==m){printf("Found at %d\n",i-j);j=lps[j-1];}
        else if(i<n&&text[i]!=pat[j]) j=j?lps[j-1]:0,i+=!j;
    }
}

File: algorithms/17_matlab/strings/kmp.m
function lps = compute_lps(pattern)
% COMPUTE_LPS Computes the LPS (longest proper prefix-suffix) array
    m = length(pattern);
    lps = zeros(1, m);
    len = 0;
    i = 2;
    while i <= m
        if pattern(i) == pattern(len + 1)
            len = len + 1;
            lps(i) = len;
            i = i + 1;
        else
            if len > 0
                len = lps(len);
            else
                lps(i) = 0;
                i = i + 1;
            end
        end
    end
end

function indices = kmp_search(text, pattern)
% KMP_SEARCH Knuth-Morris-Pratt string matching
    n = length(text);
    m = length(pattern);
    lps = compute_lps(pattern);
    i = 1;
    j = 1;
    indices = [];
    while i <= n
        if pattern(j) == text(i)
            i = i + 1;
            j = j + 1;
        end
        if j == m + 1
            indices = [indices, i - j + 1];
            j = lps(j - 1) + 1;
        elseif i <= n && pattern(j) ~= text(i)
            if j > 1
                j = lps(j - 1) + 1;
            else
                i = i + 1;
            end
        end
    end
end

function test_kmp()
    fprintf('[Matlab KMP] Testing KMP string matching\n');
    r = kmp_search('ABABDABACDABABCABAB', 'ABABCABAB');
    fprintf('Pattern found at: ');
    fprintf('%d ', r);
    fprintf('(expected 11)\n');
    r = kmp_search('AAAA', 'AA');
    fprintf('Pattern found at: ');
    fprintf('%d ', r);
    fprintf('(expected 1 2 3)\n');
    r = kmp_search('HELLO WORLD', 'WORLD');
    fprintf('Pattern found at: ');
    fprintf('%d ', r);
    fprintf('(expected 7)\n');
    fprintf('[Matlab KMP] Test completed.\n');
end